### PR TITLE
Prevent incorrect deletion of HostSubnet OVS flows

### PR DESCRIPTION
**Pull request by danwinship**
In certain circumstances we can end up with two HostSubnets having the same HostIP, which will then eventually cause nodes to delete the wrong flows:

1. node1 and node2 register with the master, the SDN master creates HostSubnets for them, and node0 observes this and creates OVS flows for them
2. The underlying VMs (or whatever) backing node1 and node2 are deleted. However, their Node objects are *not* deleted (and therefore neither are their HostSubnets).
3. node1 and node2 are recreated. Node2 comes up first, requests an IP address from its cloud/hypervisor, **and is given node1's old IP address**. It registers with the master, updating node2's Node with the new IP address. (There are now two Node objects claiming the same IP. This is apparently allowed.)
4. The SDN master sees that node2's Node's IP has changed, and changes node2's HostSubnet to match.  (There are now two HostSubnet objects claiming the same IP.)
5. node0 sees the HostSubnet update and responds by deleting its old node2 flows and creating new ones. (There are now two sets of OVS flows referring to the same node IP.)
6. Now node1 finishes booting, gets a new IP address, and updates its Node. The SDN master updates its HostSubnet. And node0 sees the update and tries to delete its old node1 flows, and accidentally ends up deleting one of the new node2 flows instead, because of the reused IP address.

This is related to #18617 but it's not the same bug; there the HostSubnet.Subnet field is changing, which would require that the HostSubnet was actually deleted and recreated. (Which means I still don't know what's really going on in #18617.)

Anyway, the fix I went with is to add a cookie based on the HostSubnet.UID to the OVS flows, and then delete based on that rather than on the HostIP/Subnet.

I don't think this problem affects any other OVS flows; pod and service IPs should actually be enforced to be unique, so this wouldn't be able to happen. But maybe we should do UID-based cookies on everything anyway?

Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1538220

### CODE REVIEW

File: pkg/network/node/ovscontroller.go [hunk 1]
```diff
@@ -503,26 +504,34 @@ func (oc *ovsController) UpdateEgressNetworkPolicyRules(policies []networkapi.Eg
...
 }
 
 func (oc *ovsController) DeleteHostSubnetRules(subnet *networkapi.HostSubnet) error {
+	cookie := hostSubnetCookie(subnet)
+
 	otx := oc.ovs.NewTransaction()
-	otx.DeleteFlows("table=10, tun_src=%s", subnet.HostIP)
-	otx.DeleteFlows("table=50, arp, nw_dst=%s", subnet.Subnet)
-	otx.DeleteFlows("table=90, ip, nw_dst=%s", subnet.Subnet)
+	otx.DeleteFlows("table=10, cookie=0x%08x", cookie)
```
Comment (pravisankar) [inline pkg/network/node/ovscontroller.go]: we could use both cookie and ip here

Comment (danwinship) [reply pkg/network/node/ovscontroller.go]: Done. I would add a test, but that would require actually coming up with a partial sha256 collision, and if I'm gonna do that I want to end up with some bitcoins at the end!

**Comment by smarterclayton** [comment]
If this is severe enough please cherry-pick to origin 3.9 once merged.

**Pull request merged**